homeassistant/components/knx/sensor.py

```diff
@@ -226,3 +263,294 @@ async def async_added_to_hass(self) -> None:
+            result["config"]["entity_config"][CONF_ENTITY_CATEGORY] = (
+                self.entity_category
+            )
+
+        # Validate before returning to set defaults
+        return self.get_schema()(result)
+
+    def to_storage_dict(self) -> dict[str, Any]:
+        """Convert the current instance into a dictionary suitable for storage."""
+
+        return {
+            CONF_GA_SENSOR: self.sensor_ga.to_dict(),
+            CONF_STATE_CLASS: self.state_class,
+            CONF_DEVICE_CLASS: self.device_class,
+            CONF_ALWAYS_CALLBACK: self.always_callback,
+            CONF_SYNC_STATE: self.sync_state,
+            CONF_NAME: self.name,
+            CONF_DEVICE_INFO: self.device_info,
+            CONF_ENTITY_CATEGORY: self.entity_category,
+        }
+
+
+class KnxUiSensorPlatformController(KnxUiEntityPlatformController):
+    """Class to manage dynamic adding and reloading of UI entities."""
+
+    async def create_entity(self, unique_id: str, config: dict[str, Any]) -> None:
+        """Add a new UI entity."""
+
+        sensorConfig = UiSensorConfig.from_storage_dict(config)
+        sensor = UiSensorEntity(self._knx_module, unique_id, sensorConfig)
+        await self._entity_platform.async_add_entities([sensor])
+
+
+class BaseSensorEntity(SensorEntity):
+    """Base class for a KNX sensor."""
+
+    _device: XknxSensor
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the state of the sensor."""
+        return self._device.resolve_state()
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return device specific state attributes."""
+        if self._device.last_telegram is not None:
+            return {
+                ATTR_SOURCE: str(self._device.last_telegram.source_address),
+            }
+        return None
+
+
+class UiSensorEntity(BaseSensorEntity, KnxUiEntity):
+    """Representation of a KNX sensor."""
+
+    _device: XknxSensor
+
+    def __init__(
+        self, knx_module: KNXModule, unique_id: str, config: SensorConfig
+    ) -> None:
+        """Initialize of a KNX sensor."""
+        super().__init__(knx_module, unique_id, {})
+
+        if config.sensor_ga.dpt is None:
+            raise ValueError("DPT must be provided for sensor entity")
+
+        dpt = DPTBase.parse_transcoder(
+            {"main": config.sensor_ga.dpt.main, "sub": config.sensor_ga.dpt.sub}
+        )
+        if dpt is None:
+            raise ValueError(f"Could not parse DPT for data: {config.sensor_ga.dpt}")
+
+        name = config.name or ""
+        always_callback = config.always_callback or False
+        sync_state = config.sync_state or True
+
+        self._device = XknxSensor(
+            xknx=knx_module.xknx,
+            name=name,
+            group_address_state=config.sensor_ga.state_ga,
+            sync_state=sync_state,
+            always_callback=always_callback,
+            value_type=dpt.value_type,
```

Comment:
Oh... an xknx.Sensor should probably accept a `type[DPTBase]` as value_type or `{main: int, sub: int | None}` dict (`xknx.dpt._DPTMainSubDict`) too. Have never noticed that it doesn't.